src/zjs_callbacks.c

```diff
@@ -315,17 +312,21 @@ void zjs_remove_callback(int32_t id)
     }
 }
 
-void zjs_signal_callback(int32_t id)
+void zjs_signal_callback(int32_t id, void* args, uint32_t size)
 {
-    if (id != -1 && cb_map[id]) {
-#ifdef DEBUG_BUILD
-        if (cb_map[id]->type == CALLBACK_TYPE_JS) {
-            DBG_PRINT("signaling JS callback id %ld\n", id);
-        } else {
-            DBG_PRINT("signaling C callback id %ld\n", id);
+    if (args && size) {
+        // signal with args
+        int ret = zjs_port_ring_buf_put(&ring_buffer, (uint16_t)id, 0, (uint32_t*)args, (uint8_t)size / 4);
+        if (ret != 0) {
+            DBG_PRINT("error putting into ring buffer, ret=%u\n", ret);
+        }
+    } else {
+        // signal without args
+        uint32_t dummy = 0;
+        int ret = zjs_port_ring_buf_put(&ring_buffer, (uint16_t)id, 0, (uint32_t*)&dummy, (uint8_t)1);
```

Comment:
So actually once you resolve this you don't need this second call to buf_put. You can just use the same code as above.

If there are no args, size is 0. size + 3 / 4 is still 0. And as my experiment above will show you, you *can* write an object to the ring buffer with size 0. It will write a 32-bit value with 16-bit "type" (id), 8-bit "value" (you've given 0), and an 8-bit "size" (0 in this case).